### PR TITLE
Initial Loki implementation of cloudsc2-nl on CPU

### PR DESCRIPTION
**Pull request by piotrows**
Current code version verifies - in addition to nl/tl/ad -  correctly on a single ATOS CPU for Loki idem and scc and scc-hoist variants. The following major code modifications present here are:
- Copy of the legacy cloudsc-loki variant in development of ml
- Inline of Cuadj* and turn cloudsc2/satur into modules, following cloudsc-loki example
- Synchronization of the CMake building strategy for Loki variant with the clouadsc-loki reference
- Global parameters are being passed via argument line rather than USE, following cloudsc reference and previous PR

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf-ifs/dwarf-p-cloudsc2-tl-ad?pullRequest=8) <br/>All committers have signed the CLA.

File: cmake/python_venv.cmake [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+##############################################################################
+#.rst:
+#
+# find_python_venv
```
Comment (mlange05) [inline cmake/python_venv.cmake]: Are we using this for the Loki setup, or is this just blindly copied? If not, please remove, as I think Loki does handle its own venvs.

File: cmake/OMP.cmake [hunk 2]
```diff
@@ -0,0 +1,44 @@
+if( HAVE_OMP )
+
+    try_compile(
```
Comment (mlange05) [inline cmake/OMP.cmake]: This OMP detection mechanism is really just for OMP device offload, right? As nothing in this PR requires it, and since we're not testing it, I'd say we can drop this until we really need it and are ready support this properly.

Comment (mlange05) [reply cmake/OMP.cmake]: Note, the same applies to the source files for the detection test.

File: src/cloudsc2_nl_loki/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,197 @@
+# Define this dwarf variant as an ECBuild feature
+ecbuild_add_option( FEATURE CLOUDSC2_NL_LOKI DEFAULT ON
+    DESCRIPTION "Build the Fortran CLOUDSC2 non-linear dwarf with Loki transformations"
+)
+
+function( cloudsc_xmod _TARGET )
+
+    if( TARGET clawfc AND ${LOKI_FRONTEND} STREQUAL "omni" )
```
Comment (mlange05) [inline src/cloudsc2_nl_loki/CMakeLists.txt]: Due to the complexity that OMNI and OFP frontends bring, I'm tempted to say we can exclude them here. We use them to test edge cases in CLOUDSC, but I think there is little gained here from carrying all of that over. Opinions? (Also paging @reuterbal ..)

Comment (mlange05) [inline src/cloudsc2_nl_loki/CMakeLists.txt]: Due to the complexity that OMNI and OFP frontends bring, I'm tempted to say we can exclude them here. We use them to test edge cases in CLOUDSC, but I think there is little gained here from carrying all of that over. Opinions? (Also paging @reuterbal ...)

File: src/cloudsc2_nl_loki/cloudsc_loki.config [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+[[routine]]
+name = 'cloudsc_driver'
+expand = true
+role = 'driver'
+
+[[routine]]
+name = 'cloudsc2'
+role = 'kernel'
+expand = true
+demote_locals = false
```
Comment (mlange05) [inline src/cloudsc2_nl_loki/cloudsc_loki.config]: Ohhhhh? :open_mouth: Is locals demotion disabled for a reason here? Does it change/impact results?

Comment (mlange05) [inline src/cloudsc2_nl_loki/cloudsc_loki.config]: Ohhhhh? :open_mouth: Is locals demotion disabled for a reason here? Does it change/impact results?

File: src/cloudsc2_nl_loki/cuadjtqs.F90 [hunk 5]
```diff
@@ -0,0 +1,281 @@
+! Copyright (C) 2003- ECMWF
+!
+! This software is licensed under the terms of the Apache Licence Version 2.0
+! which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+!
+! In applying this licence, ECMWF does not waive the privileges and immunities
+! granted to it by virtue of its status as an intergovernmental organisation
+! nor does it submit to any jurisdiction.
+!
+SUBROUTINE CUADJTQS &
```
Comment (mlange05) [inline src/cloudsc2_nl_loki/cuadjtqs.F90]: I believe this file has been manually inlined in the kernel, no? Shouldn't it also come out of the common lib, if this were not the case?

File: src/cloudsc2_nl_loki/dwarf_cloudsc.F90 [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+! Here we do bit shortcut wrt what it should be when following the model
+!VP00 = 101325._JPRB
+!ZPRES(GLOBAL_STATE%KLEV)=VP00
+!CALL GPHPRE(1,GLOBAL_STATE%KLEV,1,1,YDVAB,ZPRES,PRESF=ZPRESF)
+!DO JK=1,GLOBAL_STATE%KLEV
+!  YRECLD%CETA(JK)= ZPRESF(JK)/ZPRES(NFLEVG)
+!ENDDO
+! Instead we do this similar computation involving values already available
+DO JK=1,GLOBAL_STATE%KLEV
+  YRECLD%CETA(JK)= GLOBAL_STATE%PAP(1,JK,1)/GLOBAL_STATE%PAPH(1,GLOBAL_STATE%KLEV+1,1)
```
Comment (mlange05) [inline src/cloudsc2_nl_loki/dwarf_cloudsc.F90:103]: [optional] As a small side not, I do wonder if this ratio wouldn't be much easier to compute on-the-fly? But this does not need to be investigated as part of this PR.

File: src/common/CMakeLists.txt [hunk 7]
```diff
@@ -1,44 +1,88 @@
...
+
+
+ecbuild_add_library( TARGET cloudsc2-common-lib
+    SOURCES ${CLOUDSC2_COMMON_SOURCES}
     DEFINITIONS ${CLOUDSC_DEFINITIONS}
 )
-target_include_directories( cloudsc-common-lib PUBLIC $<INSTALL_INTERFACE:include> )
-target_include_directories( cloudsc-common-lib PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-target_include_directories( cloudsc-common-lib PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/module>)
+target_include_directories( cloudsc2-common-lib PUBLIC $<INSTALL_INTERFACE:include> )
```
Comment (mlange05) [inline src/common/CMakeLists.txt]: Could we go to the same generator-expression-based inline definition as we do in CLOUDSC? As in add this to the `ecbuild_add_library()` call above? 
```
    PRIVATE_INCLUDES
        $<${HAVE_HDF5}:${HDF5_Fortran_INCLUDE_DIRS}>
    PUBLIC_INCLUDES
        $<INSTALL_INTERFACE:include>
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
        $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/module>
```

File: src/common/CMakeLists.txt [hunk 8]
```diff
@@ -1,44 +1,88 @@
...
 if( HAVE_MPI AND TARGET MPI::MPI_Fortran )
-    target_sources( cloudsc-common-lib PRIVATE module/cloudsc_mpif.F90 )
-    target_link_libraries( cloudsc-common-lib PRIVATE MPI::MPI_Fortran )
+    target_sources( cloudsc2-common-lib PRIVATE module/cloudsc_mpif.F90 )
+    target_link_libraries( cloudsc2-common-lib PRIVATE MPI::MPI_Fortran )
+endif()
+
+
+if( HAVE_ACC )
+  # ========================================================================
```
Comment (mlange05) [inline src/common/CMakeLists.txt]: Same comment as above; we don't need to build both lib types any more. Instead we can toggle as we do in CLOUDSC.

**Review by mlange05 — CHANGES_REQUESTED** [review]
Ok, this looks nice on the whole, but I think a bunch of unnecessary boiler plate code has crept in here that isn't needed or can be updated now. I've not run this yet, but wanted to file first round early to get this going. I'll give this another go once some of the comments have been addressed.

File: src/cloudsc2_nl_loki/CMakeLists.txt (hunk 3 above)
Comment (reuterbal) [reply src/cloudsc2_nl_loki/CMakeLists.txt]: Agreed, I think we should focus on the Fparser API here and rather work on making this functional for other frontends under the hood if it is ever desired.

File: cmake/python_venv.cmake (hunk 1 above)
Comment (piotrows) [reply cmake/python_venv.cmake]: Removed

File: cmake/OMP.cmake (hunk 2 above)
Comment (piotrows) [reply cmake/OMP.cmake]: Removed

File: src/cloudsc2_nl_loki/CMakeLists.txt (hunk 3 above)
Comment (piotrows) [reply src/cloudsc2_nl_loki/CMakeLists.txt]: Omni and OFP removed

File: src/cloudsc2_nl_loki/cuadjtqs.F90 (hunk 5 above)
Comment (piotrows) [reply src/cloudsc2_nl_loki/cuadjtqs.F90]: It is inlined, I am not sure if it should come out from the common lib, it is not the case in the reference (nl variant). Removed

File: src/common/CMakeLists.txt (hunk 7 above)
Comment (piotrows) [reply src/common/CMakeLists.txt]: Done in the later PR, back ported.

File: src/common/CMakeLists.txt (hunk 8 above)
Comment (piotrows) [reply src/common/CMakeLists.txt]: Done.

File: src/cloudsc2_nl_loki/cloudsc_loki.config (hunk 4 above)
Comment (piotrows) [reply src/cloudsc2_nl_loki/cloudsc_loki.config]: removed demote=false, results on cpu are fine

**Review by mlange05 — APPROVED** [review]
Ok, the Loki-SCC/SCC-hoist variants are not yet GPU clean, and need proper disabling for CPU-only mode. I've added `DIRECTIVE none` to the CMake loki macros for now, which requires latest loki-main, but now builds clean without any artifacts on Intel/NVHPC for CPU.

I've also removed the CMake compat features for old-style OpenACC (this was for pre-3.16, we now require >=3.17) and the OpenMP device offload testing. This should be re-introduced once we actually test and support OpenMP GPU offload in this dwarf.
